internal/serviceconfig/duration.go

```diff
@@ -0,0 +1,117 @@
+/*
+ *
+ * Copyright 2023 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package serviceconfig
+
+import (
+	"encoding/json"
+	"fmt"
+	"math"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// Duration defines JSON marshal and unmarshal methods to conform to the
+// protobuf JSON spec defined [here].
+//
+// [here]: https://protobuf.dev/reference/protobuf/google.protobuf/#duration
+type Duration time.Duration
+
+func (d Duration) String() string {
+	return fmt.Sprint(time.Duration(d))
+}
+
+// MarshalJSON converts from d to a JSON string output.
+func (d Duration) MarshalJSON() ([]byte, error) {
+	ns := time.Duration(d).Nanoseconds()
+	sec := ns / int64(time.Second)
+	ns = ns % int64(time.Second)
+
+	var sign string
+	if sec < 0 {
+		sign, sec, ns = "-", -1*sec, -1*ns
+	}
```

Comment:
And so you flip the bit on ns in this case because it subtracts from the absolute value in the negative case, right? I don't know the merging of this lgoic if my above comment is right wrt correctness issue.

Reply:
True.. This needs to look at ns too.

Re: bit flipping:
If either value is non-negative, both will be non-negative.
if either is non-positive, both will be non-positive.

And the tests pass so you can trust the logic (or suggest new test cases).  :)

Reply:
You can't add them then compare whether it's less than 1, for |x| < 1, it *must* be 0 seconds and x nanoseconds. Thus, if seconds is set signs must be same.

Reply:
"If either value is non-negative, both will be non-negative.
if either is non-positive, both will be non-positive." oh is this because we don't have a uint here